### ISSUE TITLE
Alignment and padding

### ISSUE THREAD

**Issue by lassik** [pre-srfi/binary-records#3]
We should probably support:

* Packed records (no alignment anywhere)
* Standard conventions (e.g. x86-64 C ABI)
* Custom conventions

All of these are covered if the library keeps a list of the known alignment for each type. An empty list means no alignment. Standard alignment rules can be given as a pre-defined list. If users don't like them, they can make their own list.

**Comment by johnwcowan** [comment]
In [StructuresCowan](https://github.com/johnwcowan/r7rs-work/blob/master/StructuresCowan.md) you have to specify where fill bytes are to be placed explicitly, so that covers cases 1 and 3.

I don't especially want to privilege x86-64 alignment rules, though it would be straightforward to extend StructuresCowan to do so.  After all, there is also ARM and potentially other widespread architectures in future, including whatever Apple is switching to.

**Comment by mnieper** [comment]
Am So., 11. Okt. 2020 um 17:46 Uhr schrieb John Cowan <
notifications@github.com>:


> including whatever Apple is switching to.
>
> Who cares?


/ethics off

**Comment by lassik** [comment]
It's clear that we all disagree on ethics, which is the norm in any group of people. What we could do is try to work together despite our differences. Scheme will never run out of programmers who are interested in supporting commercial platforms.

**Comment by lassik** [comment]
> I don't especially want to privilege x86-64 alignment rules

Indeed, there's no reason to limit the rules to one particular architecture. A collection of arbitrary `(data-type . alignment)` pairs should suffice. If a data type is missing in the collection, assume no alignment.

Maybe the implementation should offer a preset for the host architecture's rules when those are known. Running on ARM would provide ARM presets, etc.

**Comment by mnieper** [comment]
Am So., 11. Okt. 2020 um 21:52 Uhr schrieb Lassi Kortela <
notifications@github.com>:

> It's clear that we all disagree on ethics, which is the norm in any group
> of people. What we could do is try to work together despite our
> differences. Scheme will never run out of programmers who are interested in
> supporting commercial platforms.
>
My comment was on a lighter note. ;)

**Comment by mnieper** [comment]
Back to the topic:

Instead of point 2 ("standard convention") offer "local ABI convention", which includes endianess and struct padding. This makes interfacing with local C libraries easy.

**Comment by lassik** [comment]
Excellent point. I was thinking of using this SRFI to read and write files and network traffic, which usually have the same machine-independent endianness everywhere. But using it with C FFI makes sense too.

I've been thinking of another (reasonably small and simple) SRFI for running the local C compiler from within Scheme and finding information about the C interface (things like what is `sizeof foo_t` on this machine). The local ABI alignment information could go there as well?

**Comment by mnieper** [comment]
Am Mo., 12. Okt. 2020 um 10:44 Uhr schrieb Lassi Kortela <
notifications@github.com>:

> Excellent point. I was thinking of using this SRFI to read and write files
> and network traffic, which usually have the same machine-independent
> endianness everywhere. But using it with C FFI makes sense too.
>
> I've been thinking of another (reasonably small and simple) SRFI for
> running the local C compiler from within Scheme and finding information
> about the C interface (things like what is sizeof foo_t on this machine).
> The local ABI alignment information could go there as well?
>
Using offsetof, the answer is probably yes.

But retrieving such information shouldn't go through running a C compiler à
la Autoconf at runtime. It should be built into the Scheme implementation
at the *compile-time of the Scheme implementation*.

(Any native compiler with a FFI has to know about the ABI anyway.)

Marc

**Comment by lassik** [comment]
Agreed in principle, but it may not always be that simple.

Any Scheme with a C interface has to know basic things like the size and alignment of `void*` and `int`. But what about more exotic types? Some C libraries define their own integer types and it's not possible to anticipate all of them. Implementations like Chez Scheme that load foreign libraries at runtime need to find out the type information at runtime as well.

**Comment by lassik** [comment]
Compile-via-C implementations like Gambit and Chicken can perhaps do without that information, since they can just write `foo_t` into a C source file and let the C compiler handle it, but even with these implementations, it's useful for interactive development to be able to easily query type information from the REPL.

**Comment by lassik** [comment]
If a `.so` library is loaded at runtime, I guess it would be most reliable to extract type information from the ELF/DWARF information in that file, but not sure how easy that is and how complete the information is. The DWARF debug info is commonly not available; I guess the debug info takes a lot of extra disk space for big libraries.

**Comment by mnieper** [comment]
Am Mo., 12. Okt. 2020 um 11:04 Uhr schrieb Lassi Kortela <
notifications@github.com>:

> If a .so library is loaded at runtime, I guess it would be most reliable
> to extract type information from the ELF/DWARF information in that file,
> but not sure how easy that is and how complete the information is. The
> DWARF debug info is commonly not available; I guess the debug info takes a
> lot of extra disk space for big libraries.
>
I don't think you can reliably extract such information from a binary.

**Comment by mnieper** [comment]
Am Mo., 12. Okt. 2020 um 10:59 Uhr schrieb Lassi Kortela <
notifications@github.com>:

> Agreed in principle, but it may not always be that simple.
>
> Any Scheme with a C interface has to know basic things like the size and
> alignment of void* and int. But what about more exotic types? Some C
> libraries define their own integer types and it's not possible to
> anticipate all of them. Implementations like Chez Scheme that load foreign
> libraries at runtime need to find out the type information at runtime as
> well.
>
What do you mean by exotic types?

For a struct version that uses ABI byte-order, alignment, and padding, it
is irrelevant because all the needed info is fixed by the ABI.

Extracting information from C can in fact be useful for loading libraries
whose ABI is unknown at runtime (if we really want to do this), but this
seems to be outside the scope of this SRFI idea.

**Comment by lassik** [comment]
> What do you mean by exotic types?

E.g. `libfoo` having a `foo_int_t` type. That type is an alias for one of C's built-in integer types, but we don't know which one without running a groveler.

> Extracting information from C can in fact be useful for loading libraries whose ABI is unknown at runtime (if we really want to do this), but this seems to be outside the scope of this SRFI idea.

Indeed, that belongs in the C compiler SRFI (or yet another separate SRFI). I don't have pre-SRFIs for those yet so I mentioned them in this issue tracker.

**Comment by lassik** [comment]
> whose ABI is unknown at runtime (if we really want to do this)

With load-at-runtime implementations like Chez Scheme we don't have much choice. We could write a C wrapper library in between (or use a tool like SWIG to generate one) so that the wrapper uses known sizes like uint32_t and int64_t for everything. But that's no easier than running a groveler from Scheme; indeed, SWIG runs a very sophisticated groveler internally.

**Comment by mnieper** [comment]
Am Mo., 12. Okt. 2020 um 12:27 Uhr schrieb Lassi Kortela <
notifications@github.com>:

> whose ABI is unknown at runtime (if we really want to do this)
>
> With load-at-runtime implementations like Chez Scheme we don't have much
> choice.
>
You can check the ABI at compile-time, which is better anyway because I may
only have a cross-compiler for my target system but no compiler there. (A
user running compiled Scheme code shouldn't be forced to install a C
compiler.)

> We could write a C wrapper library in between (or use a tool like SWIG to
> generate one) so that the wrapper uses known sizes like uint32_t and
> int64_t for everything. But that's no easier than running a groveler from
> Scheme; indeed, SWIG runs a very sophisticated groveler internally.
>
In general, such a wrapper is always needed as parts of a C library
functionality can be provided through macros (that look like functions),
and often is.

**Comment by lassik** [comment]
Checking the ABI at compile time would be ideal if it works. I've been trying to find solutions like that, but have always ended up with the impression that this area is not well standardized and is still quite experimental.

**Comment by lassik** [comment]
GCC and LLVM know the types internally when they compile a library like GTK so they could note down most of that info in the `.so` file itself, which would ease high-level languages' job by a pretty extremely amount, but I don't think they do. The DWARF debug info may have part of it.

**Comment by mnieper** [comment]
You can check the GNU Assembler manual to check which directives are
available in the DWARF format.

Lassi Kortela <notifications@github.com> schrieb am Mo., 12. Okt. 2020,
12:43:

**Comment by mnieper** [comment]
The binary format doesn't help when a lot of C preprocessor magic is done.
Try to use GMP as an example to see what works and what cannot.

Lassi Kortela <notifications@github.com> schrieb am Mo., 12. Okt. 2020,
12:43:
